rfc-174-consolidate-frontend-rendering-apps.md

```diff
@@ -0,0 +1,330 @@
+
+For example if a user visits a page in one app that requires the button component CSS and then moves to a page on another app with the same component, that CSS file is downloaded once per page, rather than once.
+
+We aim to build our pages and assets to be as performant as possible, but repeated downloads invalidate some of that effort. This situation happens for any repeated components across the estate, where each app will re-deliver CSS and JavaScript to the user unnecessarily.
+
+#### router / router-api are complicated and hard to maintain
+[router] routes requests for pages on the site to the correct application. `router` is complicated and hard to maintain. Moving towards a single application will remove the need for `router` entirely.
+
+`router` is a Go application that copies a database of URLs on GOV.UK from another application called [router-api]. `router` uses that database to route requests to the correct rendering application.
+
+Go is not a language that many developers on GOV.UK are comfortable with, which makes updating the application harder.
+
+The relationship between [router and router-api] is quite tangled. `router-api` is a Ruby on Rails application. The database that it shares with router is a Mongo database.  The version of Mongo we’re using is an old one, and has had to be pinned to keep [compatibility with the Mongo driver].
+
+`content-store` registers routes in `router-api`. This means that the same content data is stored in multiple places and has the potential to get out of sync.
+
+There was an attempt to switch `router-api` to [use PostgreSQL] but this was paused indefinitely due to other priorities and the complexity of the work.
+
+#### Local preview and sharing changes to GOV.UK is difficult
+Manually testing user journeys from a page is difficult, particularly where code changes are spread across more than one application. To test locally, developers have to have multiple applications running which slows down their local machines.
+
+Having fewer apps would also mean that GOV.UK Docker, our system for running apps locally, could be simplified or potentially removed entirely.
+
+Sharing changes for non-engineering disciplines to preview is also difficult. This can be done by deploying branches to integration, but these branches are frequently overwritten by dependency updates.
+
+We can use Heroku for previews but this is also complicated by multiple apps. Previewing a user journey that involves pages from more than one app cannot be done in Heroku because links from a page in one app do not link to another. For example, a journey from the homepage to a search page and then a mainstream browse page crosses multiple app boundaries, and the links would 404 in Heroku. Previews can be done using Heroku, but this would be greatly simplified by having only a single application.
+
+### `static` and `slimmer` are complicated and inflexible
```

Comment:
Should this section be the same heading level as the ones above? Currently it's `###` but the other problems above are `####`. Then I think `The system is not designed for dynamic content` and `Code locations are scattered` should be headings beneath this section?

This shows the heading structure currently:

<img width="497" alt="Screenshot 2024-06-18 at 09 25 10" src="https://github.com/alphagov/govuk-rfcs/assets/861310/6fe47e45-43e3-47bc-adaf-0b3c5cd022ff">

Reply:
You're right, and thanks for show the heading structure, that's really helpful. I think most of the headings are supposed to be a level 3, not 4. Will edit and fix up.

Reply:
@andysellick I've updated the headings to better match the heading structure what we had in the Google doc. I've updated and fixed up.

Reply:
Cool. Now looks like this, which I think is correct?

<img width="454" alt="Screenshot 2024-06-18 at 10 55 42" src="https://github.com/alphagov/govuk-rfcs/assets/861310/1c0a9a41-f274-4e87-a24c-94003126d53d">